### ISSUE TITLE
How did you visualize the  similarity matching score maps in Figure 5 of your paper?

### ISSUE THREAD

**Issue by longmalongma** [suhwan-cho/BMVOS#1]
@suhwan-cho Thank you very much for your good work. How did you visualize the  similarity matching score maps in Figure 5 of your paper?Can you share the code to visualize similarity Matching Score Maps?Thank you. My email is: dangjsh@mail2.sysu.edu.cn. Thank you very much.

![image](https://user-images.githubusercontent.com/40291722/139574382-1fb1c0d9-501d-4072-b8d3-d6374157299c.png)

**Comment by suhwan-cho** [comment]
Hi @longmalongma!

In BMVOS, the similarity matching scores are in the range of 0 to 1 theoretically, but the values are not evenly distributed in real cases. For a clear comparison, I linearly normalized the values in each score map to have the lowest value of 0 and the largest value of 1. Then, I visualized the normalized scores through heatmaps.

**Comment by longmalongma** [comment]
> Hi @longmalongma!
> 
> In BMVOS, the similarity matching scores are in the range of 0 to 1 theoretically, but the values are not evenly distributed in real cases. For a clear comparison, I linearly normalized the values in each score map to have the lowest value of 0 and the largest value of 1. Then, I visualized the normalized scores through heatmaps.

@suhwan-cho Ok, thanks for your reply, Can you provide yourcode for this visualization?

**Comment by suhwan-cho** [comment]
After saving the normalized scores in PNG format, I generated the heatmaps using the code below.


```
from PIL import Image
import numpy as np
import matplotlib.pyplot as plt

for k in [1, 2, 3, 4, 5, 6]:
    for i in [1, 4, 16, 'inf']:
        im = np.array(Image.open('../output/{}/{}.png'.format(k, i)))
        im = im[:, :, 0]
        im = np.flip(im, axis=0)

        fig = plt.pcolormesh(im, cmap='jet')
        plt.axis('off')
        fig.axes.get_xaxis().set_visible(False)
        fig.axes.get_yaxis().set_visible(False)
        plt.savefig('../output/{}/new_{}.png'.format(k, i), bbox_inches='tight', pad_inches=0)
```

**Comment by longmalongma** [comment]
> After saving the normalized scores in PNG format, I generated the heatmaps using the code below.
> 
> ```
> from PIL import Image
> import numpy as np
> import matplotlib.pyplot as plt
> 
> for k in [1, 2, 3, 4, 5, 6]:
>     for i in [1, 4, 16, 'inf']:
>         im = np.array(Image.open('../output/{}/{}.png'.format(k, i)))
>         im = im[:, :, 0]
>         im = np.flip(im, axis=0)
> 
>         fig = plt.pcolormesh(im, cmap='jet')
>         plt.axis('off')
>         fig.axes.get_xaxis().set_visible(False)
>         fig.axes.get_yaxis().set_visible(False)
>         plt.savefig('../output/{}/new_{}.png'.format(k, i), bbox_inches='tight', pad_inches=0)
> ```

@suhwan-cho Thank you very much! Similarity matching scores is a tensor, how to save similarity matching scores into PNG form.Can you provide all the codes for feature visualization?

**Comment by suhwan-cho** [comment]
I just saved it directly.
```
torchvision.utils.save_image(fg_sim, '../sim.png')
```

**Comment by longmalongma** [comment]
> I just saved it directly.
> 
> ```
> torchvision.utils.save_image(fg_sim, '../sim.png')
> ```

@suhwan-cho Ok, how to linearly normalized the values? Do you use Softmax?

**Comment by longmalongma** [comment]
> ```
> torchvision.utils.save_image(fg_sim, '../sim.png')
> ```


What are the dimensions of fg_sim?How to convert to the correct dimension?

![image](https://user-images.githubusercontent.com/40291722/139588440-a979572d-6011-4164-bccb-255bb8b45646.png)

**Comment by suhwan-cho** [comment]
Linear normalization literally means linearly normalizing the values like below.
```
A = A - min(A)
A = A / max(A)
```
The size of fg_sim is BxHWxHxW and Bx1xHxW before and after query-wise maximum operation, respectively. For a visualization, I used the latter.

**Comment by longmalongma** [comment]
> Linear normalization literally means linearly normalizing the values like below.
> 
> ```
> A = A - min(A)
> A = A / max(A)
> ```
> 
> The size of fg_sim is BxHWxHxW and Bx1xHxW before and after query-wise maximum operation, respectively. For a visualization, I used the latter.

@suhwan-cho What function do you use to reduce the HW pool to 1?

**Comment by suhwan-cho** [comment]
As I described, query-wise maximum operation.

**Comment by longmalongma** [comment]
> > Linear normalization literally means linearly normalizing the values like below.
> > ```
> > A = A - min(A)
> > A = A / max(A)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The size of fg_sim is BxHWxHxW and Bx1xHxW before and after query-wise maximum operation, respectively. For a visualization, I used the latter.
> 
> @suhwan-cho What function do you use to reduce the HW pool to 1?
@suhwan-cho What is A here?H and W?

**Comment by longmalongma** [comment]
> As I described, query-wise maximum operation.

@suhwan-cho         A = mem.max(dim=1, keepdim=True)[0]                  Is it ok?

**Comment by longmalongma** [comment]
> As I described, query-wise maximum operation.

@suhwan-cho How many epochs will it take to save the image to have the effect as shown in Figure 5?


![image](https://user-images.githubusercontent.com/40291722/139643130-13c4d54c-cbf7-4644-a9ff-ddef08cef1ae.png)

**Comment by longmalongma** [comment]
Visualization in training or visualization in testing?

**Comment by longmalongma** [comment]
> As I described, query-wise maximum operation.


Why do I generate feature visualizations that are very bad?

![image](https://user-images.githubusercontent.com/40291722/139644604-4a7d408a-6442-479a-9ebf-a23d8b78641d.png)

**Comment by suhwan-cho** [comment]
You just need to load the pre-trained model and save fg_sim as I described above. Then it will produce the similarity score maps as in the paper.

**Comment by longmalongma** [comment]
> You just need to load the pre-trained model and save fg_sim as I described above. Then it will produce the similarity score maps as in the paper.

@suhwan-cho I'm sorry that I can't achieve the effect in your paper by myself. Could you give me your complete score visualization code?thank you!

**Comment by suhwan-cho** [comment]
https://github.com/suhwan-cho/BMVOS/blob/e4de3a64894d70f6aeb6230a26ffaf17658777da/models/bmvos.py#L155-L202

Adding this code below line 198 will generate a grayscale image. 
You can obtain a heatmap image using the code I attached before.

```
tv.utils.save_image((fg_sim - torch.min(fg_sim)) / torch.max(fg_sim - torch.min(fg_sim)), '../sim.png')
```
